### DISCUSSION TITLE
Can MOOSE's data_file read data from an unstructured mesh (PiecewiseMulticonstant)?

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#18727] [Q&A General] [answered]
Hi guys,

Just wondering if the 'data_file' option from PiecewiseMulticonstant and PiecewiseMultilinear can read data from an unstructured mesh?

It seems possible from the description, but when I checked the example, it seems like it only read the **cartesian** mesh value. Most of the data_file in the examples look like this (attached below)

![image](https://user-images.githubusercontent.com/79562439/131271690-5ebebb4b-e8ee-432f-8927-a39a985ee626.png)

Just wondering, is it possible for PiecewiseMulticonstant to read unstructured grid data from a file? for example in this format >> (x,y,z,value).

Note that I tried `ReporterPointSource`, but it introduces a source term to the variable, I only want to store data from a file, so I can use it elsewhere e.g. in BCs and Materials. @lynnmunday 

Kind regards,
Traiwit

**Comment by hugary1995** [comment, marked as answer]
I think I ran into this issue before. If I remember correctly, the piecewisemultixxx readers only support data that is somehow structured due to the way it is read, i.e. x, y, z, t have to be somehow "monotone". This is required to do fast interpolation at locations not directly specified in the data file.

Imagine if you provide data in (x, y, z, value) format, that would require a projection "solve" to get the entire field.

That reporter point source is essentially relying on finite element shape functions to integrate Dirac sources. I don't think that's what you want.

**Reply by Traiwit** [reply]
Hi @hugary1995, I have a look at this: https://mooseframework.inl.gov/source/utils/MooseUtils.html

I see MOOSE has `CSVReader`, but the only examples that come with it are 'Dirac sources', just wonder if we can apply this `CSVReader` to other applications?

**Reply by hugary1995** [reply]
Yes, of course. Just include "MooseUtils.h" in the file you want to use it.
